Add list component, typed variable and synthesizer logic for these.

onionhoney opened the pull request:
I have added typed lists and integrated it into the synthesizer. To accommodate polymorphism, type variables are introduced to enable parametric types in components, which comes with unification and variable renaming. 
Also, I integrated most of the list.ml tests in PIE, and all of them are able to infer the precondition correctly with our meta-components added back. Now, empowered by typed lists, meta-components can be automatically manufactured from existing 'base' components.

SaswatPadhi commented on loopinvgen.sh:
```diff
@@ -2,9 +2,9 @@
-VERIFY="$BIN_DIR/lig-verify"
+RECORD=" $BIN_DIR/lig-record"
+INFER=" $BIN_DIR/lig-infer"
+VERIFY=" $BIN_DIR/lig-verify"
```
Why the additional spaces?

SaswatPadhi commented on build_all.sh:
```diff
@@ -20,12 +20,16 @@ if [ $? != 0 ] ; then usage ; fi
 MAKE_DEBUG=""
 
+eval $(opam env)
+
```
Let's not have opam initialization within our script. We can assume the environment to be properly setup.

onionhoney replied:
Oops sorry.

SaswatPadhi commented on build_all.sh:
```diff
@@ -20,12 +20,16 @@ if [ $? != 0 ] ; then usage ; fi
-JOBS="`cat /proc/cpuinfo | grep processor | wc -l`"
+# JOBS="`cat /proc/cpuinfo | grep processor | wc -l`"
+JOBS=4  # Core count
+
```
Did the `cpuinfo` solution not work on mac? You could replace this with: `$([[ $(uname) = 'Darwin' ]] && sysctl -n hw.logicalcpu_max || lscpu -p | egrep -v '^#' | wc -l)` (from https://stackoverflow.com/a/23378780)

onionhoney replied:
Okay I'll add these, thanks!

SaswatPadhi commented on build_all.sh:
```diff
@@ -129,4 +133,4 @@ chmod -R 777 starexec/bin
 cd starexec
-tar cvzf ../LoopInvGen_SyGuS_INV.tgz ./*
\ No newline at end of file
+tar cvzf ../LoopInvGen_SyGuS_INV.tgz ./*
```
Please configure your editor to not leave empty lines at the end of files.

SaswatPadhi commented on test_all.sh:
```diff
@@ -48,7 +48,7 @@ eval set -- "$OPTS"
   case "$1" in
     -b | --benchmarks )
-         [ -d "$2" ] || usage "Benchmarks directory [$2] not found."
+         [ -d "$2" ] || usage "Fuck you and Benchmarks directory [$2] not found."
```
That's a little too aggressive :stuck_out_tongue:

onionhoney replied:
Oh no I didn't know why I wrote this I must be drunk.

SaswatPadhi commented on src/Utils.ml:
```diff
@@ -20,6 +20,9 @@ module List = struct
     String.concat ~sep (List.map2_exn l1 l2 ~f)
 
+  let sort ~(cmp : 'a -> 'a -> int) (l : 'a list) : 'a list =
+    List.sort ~compare:cmp l
```
I'd remove this. Why not use `List.sort` directly?

onionhoney replied:
Because there was a keyword name change from ~cmp to ~compare? I might be mistaken.

SaswatPadhi commented on src/Escher_Synth.ml:
```diff
@@ -167,17 +203,24 @@ let solve_impl ?ast:(ast=false) task consts =
   bool_array.(1) <- VSet.add btrue (VSet.singleton bfalse);
+
+  (*TODO; What to do for list here ? *)
+  list_array.(1) <- VSet.singleton nil;
```
This is fine. Only `nil` should initially be present.

SaswatPadhi commented on src/Escher_Synth.ml:
```diff
@@ -154,6 +186,10 @@ let solve_impl ?ast:(ast=false) task consts =
   let bone = let vone = Th_LIA.vone
              in ((("1", (fun ars -> vone)), Const vone), Array.create ~len:vector_size vone) in
+
+  let nil = let vlist = VList ([],TVar "fresh")
```
1. Move this to `Th_List`.
2. Rename to `vempty`.
3. Use `Th_List.vempty` here.

SaswatPadhi commented on src/Th_Bool.ml:
```diff
@@ -11,10 +11,11 @@ let new_components = [
              | [Const _] -> false
              | [_] -> true
-             | _ -> false);
+             | _ -> false); *)
```
Why?!

onionhoney replied:
Sorry, forgot to reset debug changes.. Will change back!

SaswatPadhi commented on src/Th_LIA.ml:
```diff
@@ -111,4 +111,4 @@ let new_components = [
 
-let all_components = Th_Bool.all_components @ new_components
\ No newline at end of file
+let all_components = Th_Bool.all_components @ Th_List.all_components @ new_components
```
Why do we need to export `Th_List.components` here?

SaswatPadhi commented on src/Th_List.ml:
```diff
@@ -0,0 +1,22 @@
+  } ;
+]
+
+let all_components =  new_components
```
1. Please add the remaining List components from old PIE repository.
2. For now, export `Th_LIA.components` here. Using `Th_List` and choosing between LIA vs NIA would need a small change on my side. Please leave a comment that we use `Th_List` with only `Th_LIA` for now.

SaswatPadhi commented on src/Types.ml:
```diff
@@ -27,24 +29,28 @@ let to_typ (t : string) : typ =
   match (v1, v2) with
   | (VInt i1), (VInt i2) -> Int.compare i1 i2
   | (VBool b1), (VBool b2) -> Bool.compare b1 b2
+  | (VList b1), (VList b2) -> List.compare value_compare b1 b2
```
Minor, rename `b1`, `b2` to `l1`, `l2`.

SaswatPadhi commented on src/Types.ml:
```diff
@@ -58,41 +64,55 @@ let typeof (v : value) : typ =
   | VInt(i)      -> string_of_int i
   | VBool(b)     -> string_of_bool b
+  | VList(l)     -> 
+    "[" ^ List.to_string_map l ~sep:"," ~f:serialize_value ^ "]"
```
Minor: Let's use OCaml's default separator: `;`.

SaswatPadhi commented on src/Types.ml:
```diff
@@ -58,41 +64,55 @@ let typeof (v : value) : typ =
+let rec list_of_string s = 
+  let s = String.sub ~pos: 1 ~len: (String.length s - 2) s in
+  String.split ~on:',' s 
+  |> List.filter_map ~f:deserialize_value
```
1. Please move this to `Utils` as `List.of_string`
2. Replace `,` with a parameter `sep:';'`

onionhoney replied:
It uses deserialize_value which is a funciton in Types.ml though. 
To resolve this I added a second parameter to list_of_string which takes a function to fill the place of deserialize_value.

SaswatPadhi commented on src/jbuild:
```diff
@@ -46,10 +46,18 @@
+;;     (ocamlopt_flags (:standard (:include .flambda_flags)))
+;;     (libraries (core core_extended))
+;;     (preprocessor_deps ((file config.h)))
+;;     (preprocess (pps (ppx_optcomp)))))
```
Please uncomment these.

If you don't need to build these targets, you could always write `jbuild build _build/default/src/RunPIE.exe`
Although it's worth looking into your build issues with these targets.

onionhoney replied:
Yes. I have fixed it. Turned out using multiple build targets (multiple stanzas of executable) is really awkward and doesn't seem to permit sharing of modules. So I added my List_Tests into the current list of executables.

onionhoney commented:
Hi. I separated Th_List related commits from quantifier related commits. Those two commits I made only pertain to list components, list tests, test generation and the introduction of typed lists. Maybe we should approve these first, before I add the quantifiers (e.g. meta features), since those were short and based on these? 

Also I rebased so the previous commits are gone. But I replied to most of your comments and you should see I changed accordingly everywhere in my new commit. 

Last, I apologize for the sloppiness of the previous pull request. :p I'll be more carefully reviewing my code before submitting it next time.

onionhoney commented on src/Escher_Synth.ml:
```diff
@@ -113,38 +114,54 @@ let solve_impl ?ast:(ast=false) task consts =
+          | TInt -> apply_next int_array.(i)
+          | TBool -> apply_next bool_array.(i)
+          | TList t -> apply_next list_array.(i)
+            (*
```
Here I need more help. In the commented part I tried to filter from the current level of AST only those nodes whose type matches the domain of the list components. 
But for simplicity's sake I ended up doing this filtering elsewhere.

onionhoney commented on src/Escher_Synth.ml:
```diff
@@ -113,38 +114,54 @@ let solve_impl ?ast:(ast=false) task consts =
+      let typs = List.map ~f:(fun (_,x) -> typeof x.(0)) x in 
+      let codomain = 
+        let renamed_domain = rename (collect_varnames typs) c.domain in
+        let s = unify typs renamed_domain in apply s c.codomain in
```
Here unify will throw if it finds a type mismatch in the domains. That is, if the unifier fails to find a unification after renaming. Then the exception thrown is caught by the 'with' below and effectively the expansion will stop there. This is where filtering actually happens. Should we find out a way to move this up there somehow?

onionhoney commented on src/Escher_Synth.ml:
```diff
@@ -113,38 +114,54 @@ let solve_impl ?ast:(ast=false) task consts =
     let f x =
+      (* We will modify c's codomain with the union result *)
+      try 
+      let typs = List.map ~f:(fun (_,x) -> typeof x.(0)) x in 
```
Here as you can see, I get the type of the node by calculating the type of its first value on site. This can be inefficient, and i wonder if we can directly get the type of the values contained in the nodes?

onionhoney commented on src/Escher_Synth.ml:
```diff
@@ -113,38 +114,54 @@ let solve_impl ?ast:(ast=false) task consts =
+    let array_sel t = match t with 
+    | TInt -> int_array 
+    | TBool -> bool_array
+    | TList _ -> list_array 
```
Here TChar and TString is still pending. I'll add them later maybe, after all the logic is reviewed?

onionhoney commented:
@SaswatPadhi  Hi Saswat! Here are my latest changes for list components, tested and well-formatted :). Adding the meta feature is super easy so maybe we can get this approved before adding my mete feature commit. Thanks!

onionhoney commented on test/Test_Unify.ml:
```diff
@@ -0,0 +1,41 @@
+open Base
```
I can split these up into separate tests.

onionhoney commented on test/Test_Zhouheng.ml:
```diff
@@ -0,0 +1,57 @@
+open Base
```
This is my temporary tests. We can integrate them into PIE tests later on?

onionhoney commented on src/TestGen.ml:
```diff
@@ -1,11 +1,49 @@
 open Core.Quickcheck
 
-let for_type (t : Type.t) : Value.t Generator.t =
+let (--) lower upper =
```
The design of test generator is that we pass parameters along with the type, so that we can specify list length distribution and int range.

onionhoney commented on build_all.sh:
```diff
@@ -21,7 +21,7 @@ if [ $? != 0 ] ; then usage ; fi
 eval set -- "$OPTS"
 
-JOBS="`cat /proc/cpuinfo | grep processor | wc -l`"
+JOBS="$([[ $(uname) = 'Darwin' ]] && sysctl -n hw.logicalcpu_max || lscpu -p | egrep -v '^#' | wc -l)"
```
Add support for Mac OS.
